index.js

```diff
@@ -102,6 +115,54 @@ function fastifyPostgres (fastify, options, next) {
     }
   }
 
+  fastify.addHook('onRoute', routeOptions => {
+    const useTransaction = routeOptions.useTransaction || (routeOptions.options && routeOptions.options.useTransaction)
+
+    if (useTransaction) {
+      // This will rollback the transaction if the handler fails at some point
+      const onError = async (req, reply, error) => {
+        req.transactionFailed = true
+
+        try {
+          await req.pg.query('ROLLBACK')
+        } catch (err) {
+          await req.pg.query('ROLLBACK')
+        }
+      }
+
+      routeOptions.onError = addHandler(routeOptions.onError, onError)
+    }
+
+    const preHandler = async (req, reply) => {
+      const client = await pool.connect()
+      req.pg = client
```

Comment:
why are we grabbing a client from the pool before knowing if we need a transaction? or actually, don't we already know, if we're executing this code, that we need a transaction so the next conditional is unnecessary?

Reply:
OK, so we can run all of this conditionally?

I was thinking we would still need to get the client in any case, but now I see that was a mistaken thought. This also relates to https://github.com/fastify/fastify-postgres/pull/76#discussion_r577157973 where I wanted to ensutre that I always released the client

It looks like the whole lot can go in to the conditional so I'll make that change

Reply:
if the route has no option to require the use of the transaction, all the plugin code should be a no-op basically

Reply:
In order to make this work I had to re-instate these lines from the earlier attempt which ensure a client is available on the `req`:

```
  if (!fastify.hasRequestDecorator('pg')) {
    fastify.decorateRequest('pg', null)
    fastify.addHook('onRequest', async (req, reply) => {
      req.pg = fastify.pg
    })
  }
```

Reply:
doesn't look like this code is in any way related to what you're trying to do here. if you need a client, you get it from the pool, but only if you need it

Reply:
In the original issue description `pg` is added to the `req` but if we only do that conditionally in `useTransaction` then we don't have that available on the `req` anymore. This makes the API inconsistent, ie, if I change `useTransaction` from true to false I then also have to change from `req.pg` to `fastify.pg` in my code. My useTransaction=false tests failed and enabled me to catch this.